### PR TITLE
Refactor and add types

### PR DESCRIPTION
**Pull request by anxolin**
This PR does the following
- Fixes the issue where we were leaking these BIG JSON files into the client (👀 see Test section on how to check it out)
- Add strong and simpler types for both the token details (`TokenDetails`) and the token info in the list of tokens (`TokenInfo`)
- Now the parameters from the components are typed, and so it is the `getStaticProps` 🎉

Additionally, I moved for converting the logic to the `lib/tokens` (so we don't have part of the conversions/sorting in the `getStaticProps`, part in `lib`)

Lastly, I refactored a bit the components, for example, now the Listing has a `<ListItem />`

# Test

1. Open directly the Tokens page (list)

TEST 1: The pages loads. not JSON should be loaded
image

2. Click on COW token

TEST 2: A simpler token JSON and only for COW will be fetched from the network:
image

3.  HARD REFRESH (the cow details page)

Test 3: The page loads. No JSON should be 

4.  Navigate to the list (click on the header menu)

Test 4: It should load the JSON of the list. This list should be way leaner (only the fields that are needed)

image

### CODE REVIEW

File: lib/tokens.ts [hunk 1]
```diff
@@ -1,37 +1,147 @@
 import fs from 'fs'
 import path from 'path'
+import { PlatformData, Platforms, TokenDetails, TokenInfo } from 'types'
 
-const tokensPath = path.join(process.cwd(), 'data', 'tokens.json')
-const descriptionPath = path.join(process.cwd(), 'data', 'descriptions')
+const NETWORKS = ['ethereum', 'xdai']
```
Comment (anxolin) [inline lib/tokens.ts:5]: it should be now easier to add networks

File: lib/tokens.ts [hunk 2]
```diff
@@ -1,37 +1,147 @@
...
+const COW_TOKEN_ID = 'cow-protocol'
 
-export function getAllTokensData() {
-  const tokenJson = fs.readFileSync(tokensPath, 'utf-8')
-  const tokenData = JSON.parse(tokenJson)
+const TOKENS_PATH = path.join(process.cwd(), 'data', 'tokens.json')
+const DESCRIPTIONS_DIR_PATH = path.join(process.cwd(), 'data', 'descriptions')
+
+const ALL_TOKENS = _getAllTokensData()
```
Comment (anxolin) [inline lib/tokens.ts:11]: we get and prepare the data only once

File: lib/tokens.ts [hunk 3]
```diff
@@ -1,37 +1,147 @@
...
+}
+
+/**
+ *
+ * @returns All token info sorted by market cap, with COW at the top
+ */
+export function getTokensInfo(): TokenInfo[] {
+  const tokens = ALL_TOKENS.map(_toTokenInfo)
+
+  let sortedTokens = tokens.sort(_sortTokensInfoByMarketCap)
```
Comment (anxolin) [inline lib/tokens.ts:28]: sorting was moved here (instead of having it in the getStaticProps)

File: lib/tokens.ts [hunk 4]
```diff
@@ -1,37 +1,147 @@
...
+ */
+export function getTokensIds(): string[] {
+  return ALL_TOKENS.map(({ id }) => id)
+}
+
+/**
+ *
+ * @returns All token info sorted by market cap, with COW at the top
+ */
+export function getTokensInfo(): TokenInfo[] {
```
Comment (anxolin) [inline lib/tokens.ts:25]: all methods now return new typed data

**Comment by nenadV91** [comment]
@anxolin 

> TEST 1: The pages loads. not JSON should be loaded

This is what I see in incognito, is that expected?

Screenshot 2023-07-03 at 17 15 42

> TEST 2: A simpler token JSON and only for COW will be fetched from the network:

I still see tokens.json (incognito mode, hard reload)

Screenshot 2023-07-03 at 17 17 58

File: lib/tokens.ts [hunk 5]
```diff
@@ -1,37 +1,147 @@
...
+    if (platformRaw) {
+      acc[network] = _toPlatform(platformRaw)
+    }
+
+    return acc
+  }, {})
+
+  // Return the details
+  const marketData = token.market_data
+  return {
```
Comment (nenadV91) [inline lib/tokens.ts:106]: This is amazing refactor, thank you 🙏

File: lib/tokens.ts (hunk 2 above)
Comment (nenadV91) [reply lib/tokens.ts:11]: Ahh nice 👍

File: lib/tokens.ts [hunk 6]
```diff
@@ -1,37 +1,147 @@
...
+  const detailPlatforms = token.detail_platforms
+
+  const platforms = NETWORKS.reduce<Platforms>((acc, network) => {
+    const platformRaw = detailPlatforms[network]
+    if (platformRaw) {
+      acc[network] = _toPlatform(platformRaw)
+    }
+
+    return acc
+  }, {})
```
Comment (anxolin) [inline lib/tokens.ts:102]: Refactored the network support

**Review by nenadV91 — APPROVED** [review]

**Comment by anxolin** [comment]
@elena-zh @nenadV91 now is not a big deal to get the tokens.json, is just a request, it's pre-built and fast, and the content now is way smaller 49 bytes!

It is a nextjs optimisation, so when we navigate to tokens its super fast!


Look here how the requests are done when the LINKS to tokens are in the view port. This way, if I decide to go to that page, it will go extremely fast with no need to fetch anything from the network:
https://github.com/cowprotocol/cow-fi/assets/2352112/bf36717b-61c8-4659-93f1-fb06b4448e02

**Pull request merged**